core/src/main/java/org/springframework/security/authorization/method/AuthorizationManagerAfterMethodInterceptor.java

```diff
@@ -172,17 +174,30 @@ public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy strat
 		this.securityContextHolderStrategy = () -> strategy;
 	}
 
-	private void attemptAuthorization(MethodInvocation mi, Object result) {
+	public void setApplicationContext(ApplicationContext context) {
+		this.accessDeniedHandlerResolver.setApplicationContext(context);
+	}
+
+	private Object attemptAuthorization(MethodInvocation mi, Object result) {
 		this.logger.debug(LogMessage.of(() -> "Authorizing method invocation " + mi));
 		MethodInvocationResult object = new MethodInvocationResult(mi, result);
 		AuthorizationDecision decision = this.authorizationManager.check(this::getAuthentication, object);
 		this.eventPublisher.publishAuthorizationEvent(this::getAuthentication, object, decision);
 		if (decision != null && !decision.isGranted()) {
 			this.logger.debug(LogMessage.of(() -> "Failed to authorize " + mi + " with authorization manager "
 					+ this.authorizationManager + " and decision " + decision));
-			throw new AccessDeniedException("Access Denied");
+			return handleOrThrow(mi, result, decision);
 		}
 		this.logger.debug(LogMessage.of(() -> "Authorized method invocation " + mi));
+		return result;
+	}
+
+	private Object handleOrThrow(MethodInvocation mi, Object result, AuthorizationDecision decision) {
+		MethodAccessDeniedHandler deniedHandler = this.accessDeniedHandlerResolver.resolve(mi.getMethod());
+		if (deniedHandler == null) {
+			throw new AccessDeniedException("Access Denied");
+		}
+		return deniedHandler.handle(result, decision);
```

Comment:
Is there a reason to only pass the result instead of `MethodInvocationResult`? It seems a little asymmetric that an application wouldn't have access to the method invocation in both cases.